### PR TITLE
Testkit native updates

### PR DESCRIPTION
**Pull request by dmitry-timofeev**
## Overview

---
See: https://jira.bf.local/browse/ECR-3445

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

**Comment by dmitry-timofeev** [comment]
Diff (IDEA reformatted the lib.rs, I don't know if that order matters or not):

`git diff skletsun/testkit-jni-ecr-3445 testkit-native-updates --ignore-all-space src/testkit/mod.rs src/lib.rs`

```patch
diff --git a/exonum-java-binding/core/rust/src/lib.rs b/exonum-java-binding/core/rust/src/lib.rs
index e16e785c..67ea7cf2 100644
--- a/exonum-java-binding/core/rust/src/lib.rs
+++ b/exonum-java-binding/core/rust/src/lib.rs
@@ -23,12 +23,16 @@
 
 extern crate chrono;
 pub extern crate exonum;
+extern crate exonum_cli;
 #[macro_use]
 extern crate exonum_derive;
-extern crate exonum_cli;
+extern crate exonum_testkit;
+extern crate exonum_time;
 extern crate failure;
+extern crate futures;
 pub extern crate jni;
-extern crate toml;
+#[macro_use]
+extern crate lazy_static;
 #[macro_use]
 extern crate log;
 extern crate parking_lot;
@@ -37,33 +41,27 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 pub extern crate serde_json;
-
-#[macro_use]
-extern crate lazy_static;
-
-extern crate exonum_testkit;
-extern crate exonum_time;
+extern crate structopt;
 #[cfg(test)]
 extern crate tempfile;
-extern crate futures;
+extern crate toml;
 
 pub use exonum::exonum_merkledb;
+pub use jni::errors::{Error as JniError, ErrorKind as JniErrorKind, Result as JniResult};
+pub use jni::Executor;
 
-pub mod handle;
-mod proxy;
-mod proto;
-mod runtime;
-mod storage;
-mod testkit;
-pub mod utils;
-
-pub use self::handle::{as_handle, cast_handle, drop_handle, to_handle, Handle};
 pub use handle::resource_manager::*;
 pub use proxy::*;
-pub use runtime::services;
 pub use runtime::*;
 pub use storage::*;
 pub use testkit::*;
 
-pub use jni::errors::{Error as JniError, ErrorKind as JniErrorKind, Result as JniResult};
-pub use jni::Executor;
+pub use self::handle::{cast_handle, drop_handle, to_handle, Handle};
+
+pub mod handle;
+mod proto;
+mod proxy;
+mod runtime;
+mod storage;
+mod testkit;
+pub mod utils;
diff --git a/exonum-java-binding/core/rust/src/testkit/mod.rs b/exonum-java-binding/core/rust/src/testkit/mod.rs
index 3484fe32..43986852 100644
--- a/exonum-java-binding/core/rust/src/testkit/mod.rs
+++ b/exonum-java-binding/core/rust/src/testkit/mod.rs
@@ -16,27 +16,30 @@
 
 //! Provides native methods for Java TestKit support.
 
-use self::time_provider::JavaTimeProvider;
+use std::{panic, sync::Arc};
+
 use exonum::{
     blockchain::{Block, InstanceCollection, InstanceConfig},
     crypto::{PublicKey, SecretKey},
     helpers::ValidatorId,
     merkledb::BinaryValue,
-    runtime::{InstanceSpec, Runtime, RuntimeIdentifier},
+    runtime::{InstanceSpec, RuntimeIdentifier},
 };
 use exonum_testkit::{TestKit, TestKitBuilder};
 use exonum_time::{time_provider::TimeProvider, TimeServiceFactory};
-use handle::{cast_handle, drop_handle, to_handle, Handle};
 use jni::{
     objects::{JObject, JValue},
     sys::{jboolean, jbyteArray, jobjectArray, jshort},
     Executor, JNIEnv,
 };
-use std::{panic, sync::Arc};
+
+use handle::{cast_handle, drop_handle, to_handle, Handle};
 use storage::View;
 use utils::{convert_to_string, unwrap_exc_or, unwrap_exc_or_default};
 use {JavaRuntimeProxy, JniError, JniResult};
 
+use self::time_provider::JavaTimeProvider;
+
 mod time_provider;
 
 const KEYPAIR_CLASS: &str = "com/exonum/binding/common/crypto/KeyPair";
@@ -71,20 +74,15 @@ pub extern "system" fn Java_com_exonum_binding_testkit_TestKit_nativeCreateTestK
 
             let runtime =
                 JavaRuntimeProxy::new(executor.clone(), env.new_global_ref(runtime_adapter)?);
-            // TODO: remove this line after corresponding From<JavaRuntimeProxy> implementation
-            let runtime = (
-                JavaRuntimeProxy::RUNTIME_ID as u32,
-                Box::new(runtime) as Box<dyn Runtime>,
-            );
             builder = builder
-                .with_runtime(runtime)
+                .with_additional_runtime(runtime)
                 // TODO: Rewrite with protobuf: ECR-3689
                 .with_instances(instance_configs_from_java_array(&env, services)?);
 
             if let Some(instance) =
                 time_service_instance_from_java(&env, executor.clone(), time_service_spec)?
             {
-                builder = builder.with_service(instance);
+                builder = builder.with_rust_service(instance);
             }
 
             builder
```

**Comment by dmitry-timofeev** [comment]
Removed unnecessary changes.

File: exonum-java-binding/core/rust/src/testkit/mod.rs [hunk 1]
```diff
@@ -213,3 +215,142 @@ fn create_java_keypair<'a>(
...
+// Returns String value of instance's field.
+fn get_field_as_string(env: &JNIEnv, obj: JObject, field_name: &str) -> JniResult<String> {
+    convert_to_string(
+        env,
+        env.get_field(obj, field_name, "Ljava/lang/String;")?.l()?,
+    )
+}
+
+// Creates string to be parsed as `ArtifactId` from `JavaArtifactId` representation.
+fn to_artifact_id_str(java_artifact_id: String) -> String {
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/testkit/mod.rs]: Do you pass 'toString' or 'getName'?

**Comment by dmitry-timofeev** [comment]
[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project exonum-java-parent: I/O operation failed: No source found for com/exonum/binding/core/runtime/DeployArguments.java -> [Help 1]

Coveralls again — they have real I/O operation failures so often that I must have skipped this one in #1190  — a missing filter for auto-generated files.

File: exonum-java-binding/core/rust/src/testkit/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MakarovS) [reply exonum-java-binding/core/rust/src/testkit/mod.rs]: `toString`. `getName` doesn't contain runtime id.

**Comment by dmitry-timofeev** [comment]
Merged as 892596a.

**Pull request closed without merging**